File: plugins/moe/.pi/skills/extracting-requirements/scripts/aggregate_scenarios.mjs
import {
  existsSync,
  readFileSync,
  readdirSync,
  realpathSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { basename, join, resolve, sep } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import {
  compareCodePoints,
  isPlainObject,
  parseJsonLosslessly,
  pythonEqual,
  splitLines,
} from "./python_compat.mjs";

function pathParts(path, separatorPattern) {
  return path.split(separatorPattern).filter((part) => part && part !== ".");
}

function normalizeWindowsPathSpelling(path) {
  const windowsPath = path.replaceAll("/", "\\");
  const unc = windowsPath.match(/^\\\\([^\\]+)\\([^\\]+)(?:\\|$)/);
  if (unc) {
    const root = `\\\\${unc[1]}\\${unc[2]}\\`;
    const parts = pathParts(windowsPath.slice(unc[0].length), /\\+/);
    return parts.length > 0 ? `${root}${parts.join("\\")}` : root;
  }
  const drive = windowsPath.match(/^([A-Za-z]:)(\\?)/);
  if (drive) {
    const rooted = drive[2] === "\\";
    const parts = pathParts(windowsPath.slice(drive[1].length + (rooted ? 1 : 0)), /\\+/);
    if (rooted) return parts.length > 0 ? `${drive[1]}\\${parts.join("\\")}` : `${drive[1]}\\`;
    return parts.length > 0 ? `${drive[1]}${parts.join("\\")}` : drive[1];
  }
  const rooted = windowsPath.startsWith("\\");
  const parts = pathParts(windowsPath.slice(rooted ? 1 : 0), /\\+/);
  if (rooted) return parts.length > 0 ? `\\${parts.join("\\")}` : "\\";
  return parts.length > 0 ? parts.join("\\") : ".";
}

function normalizePathSpelling(path) {
  if (sep === "\\") return normalizeWindowsPathSpelling(path);
  const doubleSlashRoot = path.startsWith("//") && !path.startsWith("///");
  const root = doubleSlashRoot ? "//" : path.startsWith("/") ? "/" : "";
  const parts = pathParts(path, /\/+/);
  if (root) return parts.length > 0 ? `${root}${parts.join("/")}` : root;
  return parts.length > 0 ? parts.join("/") : ".";
}

export function loadScenarios(paths) {
  const scenarios = [];
  for (const path of paths) {
    let data;
    try {
      data = parseJsonLosslessly(readFileSync(path, "utf8"));
    } catch (error) {
      error.invalidJsonPath = path;
      throw error;
    }
    if (Array.isArray(data)) scenarios.push(...data);
    else if (isPlainObject(data) && Object.hasOwn(data, "scenarios")) {
      if (Array.isArray(data.scenarios) || typeof data.scenarios === "string") {
        scenarios.push(...data.scenarios);
      } else if (isPlainObject(data.scenarios)) {
        scenarios.push(...Object.keys(data.scenarios));
      } else {
        throw new TypeError("scenarios is not iterable");
      }
    }
  }
  return scenarios;
}

export function loadStoryTitleToId(storiesDirectory) {
  const titleMap = new Map();
  const epicFiles = readdirSync(storiesDirectory)
    .filter((name) => /^EPIC-.*\.md$/.test(name))
    .sort(compareCodePoints);

  for (const name of epicFiles) {
    let currentId;
    for (const line of splitLines(readFileSync(join(storiesDirectory, name), "utf8"))) {
      const idMatch = line.match(/^## (STORY-\p{Nd}+)/u);
      if (idMatch) currentId = idMatch[1];
      const titleMatch = line.match(/^\*\*Title:\*\* (.+)/);
      if (titleMatch && currentId) {
        titleMap.set(titleMatch[1].trim(), currentId);
        currentId = undefined;
      }
    }
  }
  return titleMap;
}

export function dedupScenarios(scenarios) {
  const seen = new Map();
  for (const scenario of scenarios) {
    if (!scenario || typeof scenario !== "object" || Array.isArray(scenario)) {
      throw new TypeError("scenario is not an object");
    }
    const title = (scenario.title ?? "").trim();
    const existing = seen.get(title);
    if (!existing) {
      seen.set(title, {
        ...scenario,
        ...(scenario.owning_story_titles
          ? { owning_story_titles: [...scenario.owning_story_titles] }
          : {}),
        ...(scenario.sources ? { sources: [...scenario.sources] } : {}),
      });
      continue;
    }

    for (const owner of scenario.owning_story_titles ?? []) {
      const owners = existing.owning_story_titles ?? [];
      if (!owners.some((value) => pythonEqual(value, owner))) {
        if (!existing.owning_story_titles) existing.owning_story_titles = [];
        existing.owning_story_titles.push(owner);
      }
    }
    for (const source of scenario.sources ?? []) {
      const sources = existing.sources ?? [];
      if (!sources.some((value) => pythonEqual(value, source))) {
        if (!existing.sources) existing.sources = [];
        existing.sources.push(source);
      }
    }
  }
  return [...seen.values()];
}

export function assignIds(scenarios) {
  let scenarioCounter = 1;
  let journeyCounter = 1;
  for (const scenario of scenarios) {
    if (scenario.kind === "journey") {
      scenario._id = `JOURNEY-${String(journeyCounter).padStart(4, "0")}`;
      journeyCounter += 1;
    } else {
      scenario._id = `SCENARIO-${String(scenarioCounter).padStart(4, "0")}`;
      scenarioCounter += 1;
    }
  }
}

export function resolveStoryRefs(scenarios, titleMap) {
  for (const scenario of scenarios) {
    scenario._owning_stories = (scenario.owning_story_titles ?? []).map((title) => {
      const storyId = titleMap.get(title.trim());
      return storyId ?? `UNRESOLVED(${title})`;
    });
  }
}

export function formatScenario(scenario) {
  const lines = [];
  const kind = scenario.kind ?? "surface";
  lines.push(`## ${scenario._id} — ${scenario.title ?? "Untitled"}`);
  lines.push("");
  if (kind === "journey") {
    lines.push("**Kind:** journey");
    lines.push("**Proof seam:** e2e");
  } else {
    lines.push(`**Kind:** ${kind}`);
    lines.push(`**Proof seam:** ${scenario.proof_seam ?? "unknown"}`);
  }
  lines.push(`**Owning stories:** ${(scenario._owning_stories ?? []).join(", ")}`);
  lines.push("");
  lines.push("**Preconditions:**");
  for (const precondition of scenario.preconditions ?? []) lines.push(`- ${precondition}`);
  lines.push("");

  if (kind === "journey") {
    lines.push("**Steps:**");
    for (const [index, step] of (scenario.steps ?? []).entries()) {
      lines.push(`${index + 1}. ${step.action ?? ""}`);
      for (const expected of step.expected ?? []) lines.push(`   → ${expected}`);
    }
    lines.push("");
    lines.push("**Final observables:**");
    for (const observable of scenario.final_observables ?? []) lines.push(`- ${observable}`);
  } else {
    lines.push("**Action:**");
    for (const step of scenario.steps ?? []) lines.push(`- ${step.action ?? ""}`);
    lines.push("");
    lines.push("**Expected observables:**");
    for (const step of scenario.steps ?? []) {
      for (const expected of step.expected ?? []) lines.push(`- ${expected}`);
    }
    for (const observable of scenario.final_observables ?? []) lines.push(`- ${observable}`);
  }
  lines.push("");
  lines.push("**Automation status:** pending");
  lines.push("**Execution command:** TBD");
  lines.push("");
  lines.push("**Sources:**");
  for (const source of scenario.sources ?? []) {
    if (isPlainObject(source)) {
      const file = source.file ?? "";
      lines.push(source.lines ? `- \`${file}:${source.lines}\`` : `- \`${file}\``);
    } else if (typeof source === "string") lines.push(`- \`${source}\``);
  }
  lines.push("");
  return lines.join("\n");
}

export function formatScenarioDocument(scenarios) {
  const journeys = scenarios.filter((scenario) => scenario.kind === "journey");
  const surfaces = scenarios.filter((scenario) => scenario.kind !== "journey");
  if (journeys.length === 0 && surfaces.length === 0) {
    return {
      content: "# Behavior Scenarios\n\nNo scenarios extracted.\n",
      journeys: 0,
      surfaces: 0,
    };
  }
  const lines = ["# Behavior Scenarios", ""];
  if (journeys.length > 0) {
    lines.push("## Journey Scenarios", "");
    for (const scenario of journeys) lines.push(formatScenario(scenario));
  }
  if (surfaces.length > 0) {
    lines.push("## Surface Scenarios", "");
    for (const scenario of surfaces) lines.push(formatScenario(scenario));
  }
  return { content: lines.join("\n"), journeys: journeys.length, surfaces: surfaces.length };
}

function usage(program) {
  return (
    `usage: ${program} [-h] -o OUTPUT --stories-dir STORIES_DIR\n` +
    "                               json_files [json_files ...]\n"
  );
}

function help(program) {
  return [
    usage(program).trimEnd(),
    "",
    "Aggregate scenarios into behavior-scenarios.md",
    "",
    "positional arguments:",
    "  json_files            Extracted scenario JSON files",
    "",
    "options:",
    "  -h, --help            show this help message and exit",
    "  -o, --output OUTPUT   Output file path",
    "  --stories-dir STORIES_DIR",
    "                        Requirements directory for resolving story title -> ID",
    "",
  ].join("\n");
}

function looksLikeNegativeNumber(value) {
  return /^-(?:\.)?\p{Nd}/u.test(value);
}

function isOptionValue(value) {
  return (
    value !== undefined &&
    (!value.startsWith("-") || value === "-" || looksLikeNegativeNumber(value))
  );
}

function parseArgs(args) {
  let output;
  let storiesDirectory;
  let optionsEnded = false;
  const jsonFiles = [];
  const unrecognized = [];
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (optionsEnded) {
      jsonFiles.push(arg);
      continue;
    }
    if (arg === "--") {
      optionsEnded = true;
    } else if (arg.startsWith("--=")) {
      return {
        error: `ambiguous option: ${arg} could match --help, --output, --stories-dir`,
      };
    } else if (arg === "-h" || (arg.startsWith("--") && "--help".startsWith(arg))) {
      return { help: true };
    } else if (
      arg === "-o" ||
      (arg.startsWith("--") && !arg.includes("=") && "--output".startsWith(arg))
    ) {
      const value = args[++index];
      if (!isOptionValue(value)) {
        return { error: "argument -o/--output: expected one argument" };
      }
      output = value;
    } else if (
      arg.startsWith("--") &&
      arg.includes("=") &&
      "--output".startsWith(arg.slice(0, arg.indexOf("=")))
    ) {
      output = arg.slice(arg.indexOf("=") + 1);
    } else if (arg.startsWith("-o") && arg.length > 2) {
      output = arg.slice(2);
    } else if (
      arg.startsWith("--") &&
      !arg.includes("=") &&
      "--stories-dir".startsWith(arg)
    ) {
      const value = args[++index];
      if (!isOptionValue(value)) {
        return { error: "argument --stories-dir: expected one argument" };
      }
      storiesDirectory = value;
    } else if (
      arg.startsWith("--") &&
      arg.includes("=") &&
      "--stories-dir".startsWith(arg.slice(0, arg.indexOf("=")))
    ) {
      storiesDirectory = arg.slice(arg.indexOf("=") + 1);
    } else if (arg.startsWith("-") && arg !== "-" && !looksLikeNegativeNumber(arg)) {
      unrecognized.push(arg);
    } else {
      jsonFiles.push(arg);
    }
  }
  if (unrecognized.length > 0) return { error: `unrecognized arguments: ${unrecognized.join(" ")}` };
  const required = [];
  if (output === undefined) required.push("-o/--output");
  if (storiesDirectory === undefined) required.push("--stories-dir");
  if (jsonFiles.length === 0) required.push("json_files");
  if (required.length > 0) {
    return { error: `the following arguments are required: ${required.join(", ")}` };
  }
  return { output, storiesDirectory, jsonFiles };
}

export async function main(args) {
  const program = basename(process.argv[1] ?? "aggregate_scenarios.mjs");
  const options = parseArgs(args);
  if (options.help) {
    process.stdout.write(help(program));
    return 0;
  }
  if (options.error) {
    process.stderr.write(`${usage(program)}${program}: error: ${options.error}\n`);
    return 2;
  }
  const jsonFiles = options.jsonFiles.map(normalizePathSpelling);
  const storiesDirectory = normalizePathSpelling(options.storiesDirectory);
  const output = normalizePathSpelling(options.output);
  for (const path of jsonFiles) {
    if (!existsSync(path)) {
      process.stderr.write(`error: file not found: ${path}\n`);
      return 2;
    }
  }
  if (!existsSync(storiesDirectory) || !statSync(storiesDirectory).isDirectory()) {
    process.stderr.write(`error: stories directory not found: ${storiesDirectory}\n`);
    return 2;
  }

  let scenarios;
  try {
    scenarios = loadScenarios(jsonFiles);
  } catch (error) {
    if (error.invalidJsonPath) {
      process.stderr.write(`error: invalid JSON in ${error.invalidJsonPath}: ${error.message}\n`);
      return 1;
    }
    throw error;
  }
  if (scenarios.length === 0) {
    process.stderr.write("warning: no scenarios found in input files\n");
    writeFileSync(output, formatScenarioDocument([]).content);
    return 0;
  }

  const deduped = dedupScenarios(scenarios);
  assignIds(deduped);
  resolveStoryRefs(deduped, loadStoryTitleToId(storiesDirectory));
  const document = formatScenarioDocument(deduped);
  writeFileSync(output, document.content);
  process.stdout.write(
    `OK: ${document.journeys} journey scenarios, ${document.surfaces} surface scenarios\n`,
  );
  return 0;
}

function isDirectEntry() {
  if (!process.argv[1]) return false;
  try {
    return realpathSync(resolve(process.argv[1])) === realpathSync(fileURLToPath(import.meta.url));
  } catch {
    return pathToFileURL(resolve(process.argv[1])).href === import.meta.url;
  }
}

const isDirect = isDirectEntry();
if (isDirect) process.exitCode = await main(process.argv.slice(2));

File: plugins/moe/.cursor-plugin/skills/extracting-requirements/scripts/aggregate_scenarios.py
#!/usr/bin/env python3
"""Aggregate extracted scenario JSONs into behavior-scenarios.md.

Usage: aggregate_scenarios.py -o <output-file> --stories-dir <requirements-dir> <json-file>...

Takes one or more JSON files (each containing {"scenarios": [...]}),
deduplicates by title, assigns stable IDs (SCENARIO-NNNN for surface,
JOURNEY-NNNN for journey), resolves owning_story_titles to STORY-IDs
using the requirements directory, and writes behavior-scenarios.md.
"""
import argparse
import json
import re
import sys
from collections import OrderedDict
from pathlib import Path


def load_scenarios(paths: list[Path]) -> list[dict]:
    """Load and combine scenarios from multiple JSON files."""
    all_scenarios: list[dict] = []
    for p in paths:
        data = json.loads(p.read_text())
        if isinstance(data, dict) and "scenarios" in data:
            all_scenarios.extend(data["scenarios"])
        elif isinstance(data, list):
            all_scenarios.extend(data)
    return all_scenarios


def load_story_title_to_id(stories_dir: Path) -> dict[str, str]:
    """Build a title -> STORY-ID map from per-epic requirement files."""
    title_map: dict[str, str] = {}
    for epic_file in sorted(stories_dir.glob("EPIC-*.md")):
        text = epic_file.read_text()
        current_id = None
        for line in text.splitlines():
            id_match = re.match(r"^## (STORY-\d+)", line)
            if id_match:
                current_id = id_match.group(1)
            title_match = re.match(r"\*\*Title:\*\* (.+)", line)
            if title_match and current_id:
                title_map[title_match.group(1).strip()] = current_id
                current_id = None
    return title_map


def dedup_scenarios(scenarios: list[dict]) -> list[dict]:
    """Deduplicate scenarios by exact title match.

    Title alone is not a safe key: two scenarios can both come back from
    extraction with a missing or empty title (e.g. an extraction subagent
    that failed to produce one), and merging them would silently drop one
    scenario's kind/preconditions/steps/final_observables and misattribute
    its citations to the survivor. An empty title is never treated as a
    match, even against another empty title.
    """
    seen: "OrderedDict[object, dict]" = OrderedDict()
    empty_title_count = 0
    for scenario in scenarios:
        title = scenario.get("title", "").strip()
        if not title:
            empty_title_count += 1
            key = ("__no_title__", empty_title_count)
        else:
            key = title
        if key not in seen:
            seen[key] = dict(scenario)
        else:
            # Merge owning_story_titles and sources
            existing = seen[key]
            for t in scenario.get("owning_story_titles", []):
                if t not in existing.get("owning_story_titles", []):
                    existing.setdefault("owning_story_titles", []).append(t)
            for s in scenario.get("sources", []):
                if s not in existing.get("sources", []):
                    existing.setdefault("sources", []).append(s)
    return list(seen.values())


def assign_ids(scenarios: list[dict]) -> None:
    """Assign stable SCENARIO-NNNN or JOURNEY-NNNN IDs."""
    scenario_counter = 1
    journey_counter = 1
    for s in scenarios:
        if s.get("kind") == "journey":
            s["_id"] = f"JOURNEY-{journey_counter:04d}"
            journey_counter += 1
        else:
            s["_id"] = f"SCENARIO-{scenario_counter:04d}"
            scenario_counter += 1


def resolve_story_refs(scenarios: list[dict], title_map: dict[str, str]) -> None:
    """Replace owning_story_titles with resolved STORY-IDs where possible."""
    for s in scenarios:
        resolved = []
        for title in s.get("owning_story_titles", []):
            story_id = title_map.get(title.strip())
            if story_id:
                resolved.append(story_id)
            else:
                resolved.append(f"UNRESOLVED({title})")
        s["_owning_stories"] = resolved


def format_scenario(s: dict) -> str:
    """Format one scenario as markdown."""
    lines: list[str] = []
    sid = s["_id"]
    title = s.get("title", "Untitled")
    lines.append(f"## {sid} — {title}")
    lines.append("")

    kind = s.get("kind", "surface")
    seam = s.get("proof_seam", "unknown")
    stories = ", ".join(s.get("_owning_stories", []))

    if kind == "journey":
        lines.append(f"**Kind:** journey")
        lines.append(f"**Proof seam:** e2e")
    else:
        lines.append(f"**Kind:** {kind}")
        lines.append(f"**Proof seam:** {seam}")

    lines.append(f"**Owning stories:** {stories}")
    lines.append("")

    lines.append("**Preconditions:**")
    for p in s.get("preconditions", []):
        lines.append(f"- {p}")
    lines.append("")

    if kind == "journey":
        lines.append("**Steps:**")
        for i, step in enumerate(s.get("steps", []), 1):
            action = step.get("action", "")
            lines.append(f"{i}. {action}")
            for exp in step.get("expected", []):
                lines.append(f"   → {exp}")
        lines.append("")
        lines.append("**Final observables:**")
        for obs in s.get("final_observables", []):
            lines.append(f"- {obs}")
    else:
        lines.append("**Action:**")
        for step in s.get("steps", []):
            lines.append(f"- {step.get('action', '')}")
        lines.append("")
        lines.append("**Expected observables:**")
        for step in s.get("steps", []):
            for exp in step.get("expected", []):
                lines.append(f"- {exp}")
        for obs in s.get("final_observables", []):
            lines.append(f"- {obs}")
    lines.append("")

    lines.append("**Automation status:** pending")
    lines.append("**Execution command:** TBD")
    lines.append("")

    lines.append("**Sources:**")
    for src in s.get("sources", []):
        if isinstance(src, dict):
            f = src.get("file", "")
            l = src.get("lines", "")
            lines.append(f"- `{f}:{l}`" if l else f"- `{f}`")
        elif isinstance(src, str):
            lines.append(f"- `{src}`")
    lines.append("")

    return "\n".join(lines)


def main() -> int:
    parser = argparse.ArgumentParser(description="Aggregate scenarios into behavior-scenarios.md")
    parser.add_argument("-o", "--output", required=True, help="Output file path")
    parser.add_argument("--stories-dir", required=True,
                        help="Requirements directory for resolving story title -> ID")
    parser.add_argument("json_files", nargs="+", help="Extracted scenario JSON files")
    args = parser.parse_args()

    paths = [Path(p) for p in args.json_files]
    for p in paths:
        if not p.exists():
            print(f"error: file not found: {p}", file=sys.stderr)
            return 2

    stories_dir = Path(args.stories_dir)
    if not stories_dir.is_dir():
        print(f"error: stories directory not found: {stories_dir}", file=sys.stderr)
        return 2

    scenarios = load_scenarios(paths)
    if not scenarios:
        print("warning: no scenarios found in input files", file=sys.stderr)
        Path(args.output).write_text("# Behavior Scenarios\n\nNo scenarios extracted.\n")
        return 0

    deduped = dedup_scenarios(scenarios)
    assign_ids(deduped)

    title_map = load_story_title_to_id(stories_dir)
    resolve_story_refs(deduped, title_map)

    # Separate journeys and surface scenarios
    journeys = [s for s in deduped if s.get("kind") == "journey"]
    surfaces = [s for s in deduped if s.get("kind") != "journey"]

    lines: list[str] = ["# Behavior Scenarios", ""]

    if journeys:
        lines.append("## Journey Scenarios")
        lines.append("")
        for s in journeys:
            lines.append(format_scenario(s))

    if surfaces:
        lines.append("## Surface Scenarios")
        lines.append("")
        for s in surfaces:
            lines.append(format_scenario(s))

    Path(args.output).write_text("\n".join(lines))

    print(f"OK: {len(journeys)} journey scenarios, {len(surfaces)} surface scenarios")
    return 0


if __name__ == "__main__":
    sys.exit(main())
